### ISSUE TITLE
Clear struct array grad keepalive on replacement

### ISSUE THREAD

**Issue by shi-eric** [NVIDIA/warp#1520]
## Summary

Plain `wp.array` fields on Warp structs retain a `_field_grad` keepalive when assigned a `requires_grad=True` array. That keepalive is needed if users later toggle `struct.field.requires_grad = False`, because the struct setter does not run and the packed C field can still hold the old gradient pointer.

However, the plain array setter does not clear the keepalive when the field is replaced with `None` or a non-grad array. In those cases the setter *does* run and rewrites the struct C field, so the old gradient array is no longer semantically referenced but remains strongly retained by `_field_grad`.

The indexed-array setter already handles this by clearing its keepalive whenever the replacement value cannot reference a grad array.

## Repro

```python
import gc
import weakref

import warp as wp


@wp.struct
class StructWithArray:
    data: wp.array(dtype=float)


wp.init()

for replacement in ("none", "non_grad_array"):
    s = StructWithArray()
    old_array = wp.array([1.0, 2.0, 3.0], dtype=float, device="cpu", requires_grad=True)
    old_grad_ref = weakref.ref(old_array.grad)

    s.data = old_array
    del old_array

    if replacement == "none":
        s.data = None
    else:
        s.data = wp.array([4.0, 5.0, 6.0], dtype=float, device="cpu")

    gc.collect()

    print(replacement, getattr(s, "_data_grad", None), old_grad_ref())
    assert getattr(s, "_data_grad", None) is None
    assert old_grad_ref() is None
```

## Actual behavior

Both replacement paths leave `s._data_grad` pointing at the previous gradient array, so `old_grad_ref()` remains live after `gc.collect()`.

## Expected behavior

After assigning `None` or a non-grad array to the struct field, the stale `_data_grad` keepalive should be cleared and the previous gradient array should be collectible when no other references remain.

## Notes

The likely fix is to mirror the indexed-array setter behavior in the plain array setter: set `_field_grad` when the new value requires grad, otherwise clear it. The existing `requires_grad` toggle behavior should continue retaining the grad keepalive because no setter runs in that path.

**Comment by shi-eric** [comment]
This is addressed by 84bb22c3b647dd2d2d8eefd9fc63b98afeb8c574.

The change updates the plain `wp.array` struct field setter so stale `_<field>_grad` keepalives are cleared when the field is replaced with `None` or a non-gradient array. It preserves the existing requires-grad toggle behavior because that path does not rerun the setter.

Test coverage:
- New tests in `warp/tests/test_struct.py`: `test_struct_array_gc_replacement_clears_grad_keepalive` covers both replacement paths and verifies that `_data_grad` is cleared and the previous gradient array can be collected.

This issue can be closed as completed.
